scapy/contrib/ethercat.uts

```diff
@@ -0,0 +1,269 @@
+    bf = BitFieldUserExample(**data)
+    # rebuild big-endian and little-endian bitfields from its own binary expressions
+    bf_le = BitFieldUserExampleLE(bf_le.do_build())
+    bf = BitFieldUserExample(bf.do_build())
+    ''' disabled as only required for 'visual debugging'
+    from scapy.compat import raw
+    # dump content for debugging
+    bitstr = ''
+    hexstr = ''
+    for i in bytearray(raw(bf)):
+        bitstr += '{:08b} '.format(i)
+        hexstr += '{:02x} '.format(i)
+    print('BE - BITS: {} HEX: {}  ({})'.format(bitstr, hexstr, data))
+    bitstr = ''
+    hexstr = ''
+    for i in bytearray(raw(bf_le)):
+        bitstr += '{:08b} '.format(i)
+        hexstr += '{:02x} '.format(i)
+    print('LE - BITS: {} HEX: {}  ({})'.format(bitstr, hexstr, data))
+    '''
+    # compare values
+    for key in data:
+        assert(getattr(bf,key) == data[key])
+        assert (getattr(bf_le, key) == data[key])
+
+= Avoid mix of LEBitFields and BitFields
+
+TEST_SAMPLE_ENUM = {
+    0x01: 'one',
+    0x02: 'two',
+    0x03: 'three',
+    0x04: 'four',
+    0x05: 'five',
+    0x06: 'six',
+    0x07: 'seven'
+}
+
+class MissingFieldSameLEFieldTypes(Packet):
+
+    fields_desc = [
+        LEBitEnumField('a', 0, 2, TEST_SAMPLE_ENUM),
+        LEBitField('b', 0, 18),
+    ]
+
+try:
+    frm = MissingFieldSameLEFieldTypes().build()
+    assert False
+except LEBitFieldSequenceException:
+    pass
+
+
+class MissingFieldDifferentLEFieldTypes(Packet):
+
+    fields_desc = [
+        LEBitEnumField('a', 0, 2, TEST_SAMPLE_ENUM),
+        LEBitField('b', 0, 18),
+    ]
+
+try:
+    frm = MissingFieldDifferentLEFieldTypes().build()
+    assert False
+except LEBitFieldSequenceException:
+    pass
+
+
+class MixedBitFieldTypesLEBE(Packet):
+
+    fields_desc = [
+        LEBitField('a', 0, 12),
+        BitField('b', 0, 4),
+    ]
+
+try:
+    frm = MixedBitFieldTypesLEBE().build()
+    assert False
+except LEBitFieldSequenceException:
+    pass
+
+
+class MixedBitFieldTypesBELE(Packet):
+
+    fields_desc = [
+        BitField('b', 0, 4),
+        LEBitField('a', 0, 12),
+    ]
+
+try:
+    frm = MixedBitFieldTypesBELE().build()
+    assert False
+except LEBitFieldSequenceException:
+    pass
+
+################################################
++ EtherCat header layer handling
+= EtherCat and padding
+
+frm = Ether() / EtherCat()
+# even with padding the length must be zero
+# the Ether(do_build()) forces the calculation of all (post_build generated) fields
+frm = Ether(frm.do_build())
+assert(frm[EtherCat].length == 0)
+assert(len(frm) == 60)
+frm = Ether()/Dot1Q()/Dot1Q()/EtherCat()
+frm = Ether()/EtherCat()
+assert(len(frm) == 60)
+frm = Ether(frm.do_build())
+assert(frm[EtherCat].length == 0)
+
+= EtherCat and RawPayload
+
+frm=Ether()/EtherCat()/Raw('0123456789')
+assert(len(frm) == 60)
+frm = Ether(frm.do_build())
+assert(frm[EtherCat].length == 10)
+frm = Ether()/EtherCat()/Raw('012345678901234567890123456789012345678901234567890123456789')
```

Comment:
You need to write:

`frm = Ether()/EtherCat()/Raw(b'012345678901234567890123456789012345678901234567890123456789')
`

The same comment apply to other part of your code.